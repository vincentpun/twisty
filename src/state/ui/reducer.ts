import { combineReducers } from 'redux';

import appControl, { AppControlUIState } from './AppControl/reducer';

export interface UIState {
  appControl: Readonly<AppControlUIState>,
}

export default combineReducers<UIState>({
  appControl,
});
