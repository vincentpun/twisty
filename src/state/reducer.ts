import { combineReducers } from 'redux';

import ui, { UIState } from './ui/reducer';
import routes, { RoutesState } from './routes/reducer';

export interface State {
  ui: Readonly<UIState>
  routes: Readonly<RoutesState>,
}

export default combineReducers<State>({
  ui,
  routes,
});
