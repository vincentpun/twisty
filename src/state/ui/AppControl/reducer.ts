import { combineReducers } from 'redux';

import { createAction } from 'src/state/actions';
import { AppControlUIActionTypes } from './actionTypes';
import { AppControlSelection } from './types';
import { Actions } from './actions';

export interface AppControlUIState {
  selection: AppControlSelection;
}

const initialState: AppControlUIState = {
  selection: null,
};

export const selection = (state = initialState.selection, action: Actions) => {
  switch (action.type) {
    case AppControlUIActionTypes.Select:
      return action.payload;
    default:
      return state;
  }
}

export default combineReducers<AppControlUIState>({
  selection,
});
