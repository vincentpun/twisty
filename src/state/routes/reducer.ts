import { combineReducers } from 'redux';

import { Actions } from './actions';
import { RoutesActionsTypes } from './actionTypes';
import { MapCoordinates } from '../ui/AppControl/types';

export enum RouteQueryStatus {
  NotStarted = 'ROUTE_QUERY_STATUS_NOT_STARTED',
  FetchingRoutesToken = 'ROUTE_QUERY_STATUS_FETCHING_ROUTES_TOKEN',
  ToPoll = 'ROUTE_QUERY_STATUS_TO_POLL',
  PollingInProgress = 'ROUTE_QUERY_STATUS_POLLING_IN_PROGRESS',
  Success = 'ROUTE_QUERY_STATUS_SUCCESS',
  Error = 'ROUTE_QUERY_STATUS_ERROR',
}

export interface RoutesState {
  queries: {
    [id: string]: RouteQuery;
  };
}

export interface RouteQuery {
  id: string;
  status: RouteQueryStatus;
  token?: string;
  response?: {
    waypoints: MapCoordinates[];
    totalDistance: number;
    totalTime: number;
    [key: string]: any;
  }
  error?: any;
}

const initialState = {
  queries: {} as { [id: string]: RouteQuery },
};

export const queries = (state = initialState.queries, action: Actions) => {
  switch (action.type) {
    case RoutesActionsTypes.CreateQuery: {
      const id = action.payload;
      return {
        ...state,
        [id]: {
          id,
          status: RouteQueryStatus.NotStarted,
        },
      };
    }
    case RoutesActionsTypes.FetchRoutesToken: {
      const id = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.FetchingRoutesToken,
        },
      };
    }
    case RoutesActionsTypes.FetchRoutesTokenReceive: {
      const { id, token } = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.ToPoll,
          token,
        },
      };
    };
    case RoutesActionsTypes.PollResult: {
      const id = action.payload;

      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.PollingInProgress,
        },
      };
    }
    case RoutesActionsTypes.PollResultReceiveInProgress: {
      const id = action.payload;

      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.ToPoll,
        },
      };
    }
    case RoutesActionsTypes.QuerySuccess: {
      const id = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.Success,
        },
      };
    }
    case RoutesActionsTypes.QueryReceive: {
      const { id, response } = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          response,
        }
      };
    }
    case RoutesActionsTypes.QueryError: {
      const { id, error } = action.payload;
      return {
        ...state,
        [id]: {
          ...state[id],
          status: RouteQueryStatus.Error,
          error,
        }
      };
    }
    default:
      return state;
  }
}

export default combineReducers<RoutesState>({
  queries,
});
