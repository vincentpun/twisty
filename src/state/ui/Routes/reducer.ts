import { combineReducers } from 'redux';
import { Actions } from 'src/state/routes/actions';
import { RoutesActionsTypes } from '../../routes/actionTypes';

export interface RoutesState {
  query: Readonly<string>;
}

export const query = (state: string = null, action: Actions) => {
  switch (action.type) {
    case RoutesActionsTypes.CreateQuery:
      return action.payload;
    default:
      return state;
  }
};

export default combineReducers({
  query,
});
