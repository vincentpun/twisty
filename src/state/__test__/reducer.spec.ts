import reducer from '../reducer';

describe('state', () => {
  it('matches the snapshot', () => {
    const state = reducer(undefined, { type: null });
    expect(state).toMatchSnapshot();
  });
});
