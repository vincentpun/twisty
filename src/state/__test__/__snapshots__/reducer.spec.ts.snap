// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`state matches the snapshot 1`] = `
Object {
  "routes": Object {
    "queries": Object {},
  },
  "ui": Object {
    "appControl": Object {
      "coordinates": Object {},
      "dropoffCoordinates": Array [],
      "hasFailedGrabbingClientLocation": false,
      "isGrabbingClientLocation": false,
      "selection": null,
      "startingCoordinates": null,
    },
    "mode": "appControl",
    "routes": Object {
      "query": null,
    },
  },
}
`;
